src/dolphin/goldstein.py

```diff
@@ -0,0 +1,97 @@
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+def goldstein(phase: ArrayLike, alpha: float, psize: int = 32) -> np.ndarray:
+    """Apply the Goldstein adaptive filter to the given data.
+
+    Parameters
+    ----------
+    phase : np.ndarray
+        2D complex array containing the data to be filtered.
+    alpha : float
+        Filtering parameter for Goldstein algorithm
+        Must be between 0 (no filtering) and 1 (maximum filtering)
+    psize : int, optional
+        edge length of square patch
+        Default = 32
+
+    Returns
+    -------
+        2D numpy array of filtered data.
+
+    """
+
+    def apply_pspec(data: ArrayLike):
+        # NaN is allowed value
+        assert not (alpha < 0), f"Invalid parameter value {alpha} < 0"
+        wgt = np.power(np.abs(data) ** 2, alpha / 2)
+        data = wgt * data
+        return data
+
+    def make_wgt(nxp: int, nyp: int) -> np.ndarray:
+        # Create arrays of horizontal and vertical weights
+        wx = 1.0 - np.abs(np.arange(nxp // 2) - (nxp / 2.0 - 1.0)) / (nxp / 2.0 - 1.0)
+        wy = 1.0 - np.abs(np.arange(nyp // 2) - (nyp / 2.0 - 1.0)) / (nyp / 2.0 - 1.0)
+        # Compute the outer product of wx and wy to create
+        # the top-left quadrant of the weight matrix
+        quadrant = np.outer(wy, wx)
+        # Create a full weight matrix by mirroring the quadrant along both axes
+        wgt = np.block(
+            [
+                [quadrant, np.flip(quadrant, axis=1)],
+                [np.flip(quadrant, axis=0), np.flip(np.flip(quadrant, axis=0), axis=1)],
+            ]
+        )
+        return wgt
+
+    def patch_goldstein_filter(
+        data: ArrayLike, wgt: ArrayLike, psize: int
+    ) -> np.ndarray:
+        """Apply the filter to a single patch of data.
+
+        Parameters
+        ----------
+        data : np.ndarray
+            2D complex array containing the data to be filtered.
+        wgt : np.ndarray
+            weight matrix for summing neighboring data
+        psize : int
+            edge length of square FFT area
+
+        Returns
+        -------
+            2D numpy array of filtered data.
+
+        """
+        # Calculate alpha
+        data = np.fft.fft2(data, s=(psize, psize))
+        data = apply_pspec(data)
+        data = np.fft.ifft2(data, s=(psize, psize))
+        return wgt * data
+
+    def apply_goldstein_filter(data: ArrayLike) -> np.ndarray:
+        # Create an empty array for the output
+        out = np.zeros(data.shape, dtype=np.complex64)
+        # ignore processing for empty chunks
+        if np.all(np.isnan(data)):
+            return data
```

Comment:
This could be a good thing to test- For the stitched, geocoded interferograms, The tilt means it's about 30% nodata. For single bursts, it's about 60% nodata (which is why adding a check like this speeds up the wrapped phase part by ~2x).

On the other hand, having this hardcoded is very geocoded-input-centric. An RIFG has no need for this and only gets it slowed down

Reply:
Yeah will remove this. I think this is a remnant from when I was nanning out zero values, which can't be done with tophu at the moment anyway.

Reply:
>For the stitched, geocoded interferograms, The tilt means it's about 30% nodata. For single bursts, it's about 60% nodata (which is why adding a check like this speeds up the wrapped phase part by ~2x).

Right now, the check for all NaN values is outside the loop over patches. So this shortcut will only trigger if the dataset contains 100% nodata. IIUC, there shouldn't be any speed up for interferograms that contain 60% nodata.

Maybe we should consider adding a similar check inside the loop over patches? If an individual patch contains all NaNs, there's no need to compute & apply the filter there.

Reply:
oh right, yes this should be removed then, the comment `# ignore processing for empty chunks` fooled me